### PR TITLE
Added the Sponsorship model

### PR DESCRIPTION
**Pull request by narayanaditya95**
@gvwilson @pbanaszkiewicz 
Writing tests. Ready for review.

### CODE REVIEW

File: workshops/models.py [hunk 1]
```diff
@@ -159,6 +159,35 @@ def self_organized_workshops_per_year_remaining(self):
 #------------------------------------------------------------
 
 
+class Sponsorship(models.Model):
+    '''Represent sponsorship from a host for an event.'''
+
+    SPONSORSHIP_SLABS = (
```
Comment (narayanaditya95) [inline workshops/models.py]: @gvwilson @pbanaszkiewicz I would like to know whether a choice field would be better suited for SWC/DC workshops or a plain simple `IntegerField`?
If we go with choices, what are the slabs in which SWC/DC accept sponsorship.

File: workshops/static/css/amy.css [hunk 2]
```diff
@@ -93,10 +93,9 @@ body {
 
 /* when using nav-tab (http://getbootstrap.com/javascript/#tabs) move the
    crispy form down a few notches */
-div[role="tabpanel"] form {
+.tab-content {
```
Comment (narayanaditya95) [inline workshops/static/css/amy.css]: This CSS selector is more specific.
`div[role="tabpanel"] form` also added padding above the `Delete` buttons in the sponsorship table.

File: workshops/models.py (hunk 1 above)
Comment (pbanaszkiewicz) [inline workshops/models.py]: For me slab is [this](http://peppardbuildingsupplies.com/sites/default/files/SWATCH_BS_Grey_0.jpg) :-)

Anyway, what does `Sponsorship.slab` represent here?

File: workshops/models.py [hunk 3]
```diff
@@ -159,6 +159,35 @@ def self_organized_workshops_per_year_remaining(self):
...
+        ('slab_a', 'Slab A'),
+        ('slab_b', 'Slab B'),
+    )
+
+    organization = models.ForeignKey(
+        'Host',
```
Comment (pbanaszkiewicz) [inline workshops/models.py]: Switch to `Organization` after #849 is merged.

File: workshops/models.py (hunk 1 above)
Comment (narayanaditya95) [inline workshops/models.py]: I am not sure how I ended up with such a field name.
Anyway, a `slab` is intended to show the divisions in which SWC/DC accept sponsorships. Each slab would have predefined sponsorship amount.
PyData uses it and it's sponsors can opt for divisions such as `PyData Founding`, `PyData Gold`, `PyData Silver`... each with decreasing sponsor amount and benefits.

File: workshops/models.py [hunk 4]
```diff
@@ -726,6 +755,10 @@ class Event(AssignmentMixin, models.Model):
         on_delete=models.PROTECT,
         help_text='Organization responsible for administrative work.'
     )
+    sponsors = models.ManyToManyField(
+        Host, related_name='sponsor', blank=True,
```
Comment (pbanaszkiewicz) [inline workshops/models.py]: Am I reading this correctly? `Event.sponsors → Hosts` but `Host.sponsor → Event`?

File: workshops/models.py (hunk 3 above)
Comment (narayanaditya95) [inline workshops/models.py]: Yes.

File: workshops/views.py [hunk 5]
```diff
@@ -156,6 +157,22 @@ def get_success_message(self, cleaned_data):
...
+
+    def delete(self, request, *args, **kwargs):
+        '''Workaround for https://code.djangoproject.com/ticket/21926'''
+        messages.success(
+            self.request,
+            self.success_message.format(self.get_object())
+        )
+        return super().delete(request, *args, **kwargs)
+
+
```
Comment (pbanaszkiewicz) [inline workshops/views.py]: You sure you need this? We used to redirect right after deletion.

File: workshops/models.py (hunk 4 above)
Comment (narayanaditya95) [inline workshops/models.py]: Yes, just as `host.administrator` gives the list of events that the host administered, `host.sponsor` gives the list of events sponsored.
On a second thought, I think `Event.sponsor_set` sounds better.

File: workshops/views.py (hunk 5 above)
Comment (narayanaditya95) [inline workshops/views.py]: Yes, we currently also redirect right after deletion. There is no intermediate confirmation page (except for the confirm javascript popup).

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 looks OK.

I wonder what we should use for the sponsorship levels aka. slabs. You can imagine we don't have an established values for them in SWC/DC. Maybe @gvwilson can provide some more feedback here?

File: workshops/models.py (hunk 3 above)
Comment (pbanaszkiewicz) [inline workshops/models.py]: That was just a note for future :-)

File: workshops/models.py (hunk 4 above)
Comment (pbanaszkiewicz) [inline workshops/models.py]: Ugh, both `organization.administrator` and `organization.sponsor` sound bad, and I believe I'm responsible for the former :-)

I'd use `organization.administered_events` or `organization.administered_set` and `organization.sponsored_events` or `organization.sponsored_set`.

File: workshops/views.py (hunk 5 above)
Comment (pbanaszkiewicz) [inline workshops/views.py]: So why do we need this view?

**Comment by narayanaditya95** [comment]
If SWC/DC does not use a predefined levels, we can switch to an `IntegerField`.
PyData package would then monkeypatch the `slab` field to add `choices` to it. Each predefined level at PyData has a sponsorship amount against it.

File: workshops/views.py (hunk 5 above)
Comment (narayanaditya95) [inline workshops/views.py]: Because it makes the task of deleting and adding a success message easier and non-repetitive.

File: workshops/models.py [hunk 6]
```diff
@@ -159,6 +159,35 @@ def self_organized_workshops_per_year_remaining(self):
...
+        help_text='Organization sponsoring the event'
+    )
+    event = models.ForeignKey(
+        'Event',
+        on_delete=models.CASCADE,
+    )
+    level = models.PositiveIntegerField(
+        choices=SPONSORSHIP_LEVELS,
+        default=1000,
+    )
```
Comment (narayanaditya95) [inline workshops/models.py]: @gvwilson Does SWC/DC use predefined levels of sponsorship, or should we simply remove the choices?

Comment (gvwilson) [inline workshops/models.py]: Our current levels are defined on 
http://software-carpentry.org/scf/join/.  I think a Partnership model 
with names is overkill - just a text field naming the level should be 
enough, and we can add special notes to the Notes field on the 
organization where needed.

Comment (narayanaditya95) [inline workshops/models.py]: The levels defined at http://software-carpentry.org/scf/join/ list the kinds of partnership between an organization and SWC/DC. It does not list the kind of sponsorship an organization can provide for an SWC/DC workshop.
What practice does SWC/DC follow while accepting sponsorship for a particular workshop?

Comment (gvwilson) [inline workshops/models.py]: My apologies, I misunderstood the question - organizational sponsorship 
for a specific workshop varies so widely that I don't think it's worth 
trying to capture as anything except a free-text field (in Notes).

**Comment by narayanaditya95** [comment]
@gvwilson I added a nullable `DecimalField` that would allow SWC/DC admins to record monetary sponsorship. The `pydata` app would on the other hand, extend the `DecimalField` to add predefined levels.

@pbanaszkiewicz I have added tests and addressed all the comments raised on the PR.
It is up for review and merge.

**Comment by narayanaditya95** [comment]
bump.

File: workshops/views.py (hunk 5 above)
Comment (pbanaszkiewicz) [inline workshops/views.py]: That's a nice change, can you do that for other deletion views if feasible (can be in a different PR)?

File: workshops/templates/workshops/event_edit_form.html [hunk 7]
```diff
@@ -4,32 +4,68 @@
...
-  </table>
-  {% else %}
-  <p>No tasks.</p>
-  {% endif %}
-
+<div>
+  <ul class="nav nav-tabs" role="tablist">
+    <li role="presentation" class="active"><a href="#event" aria-controls="event" role="tab" data-toggle="tab">Event</a></li>
+    <li role="presentation"><a href="#tasks" aria-controls="tasks" role="tab" data-toggle="tab">Tasks</a></li>
+    <li role="presentation"><a href="#sponsors" aria-controls="sponsors" role="tab" data-toggle="tab">Sponsors</a></li>
```
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/event_edit_form.html]: See how `person-edit` page behaves after you add a new task to the person -- it switches you to a tab you just were at.

You'll receive 1000 bonus virtual points if you make the tabs switch depending on the anchor in the browser's address bar. :-)

File: workshops/views.py [hunk 8]
```diff
@@ -1194,7 +1214,20 @@ def event_edit(request, event_ident):
...
                 messages.error(request, 'Fix errors below.')
+        if 'sponsor-event' in request.POST:
+            sponsor_form = SponsorshipForm(request.POST, prefix='sponsor')
 
+            if sponsor_form.is_valid():
+                sponsor = sponsor_form.save()
+
+                messages.success(
+                    request,
+                    '{} was added a new sponsor.'.format(sponsor.organization)
```
Comment (pbanaszkiewicz) [inline workshops/views.py]: Should there be "as" instead of "a"?

File: workshops/templates/workshops/event_edit_form.html (hunk 7 above)
Comment (pbanaszkiewicz) [inline workshops/templates/workshops/event_edit_form.html]: PS the trick is to use additional tags with the `message.success`. See `person-edit` for details.

File: workshops/views.py [hunk 9]
```diff
@@ -1194,7 +1214,20 @@ def event_edit(request, event_ident):
...
+                )
+                # to reset the form values
+                return redirect(request.path)
```
Comment (pbanaszkiewicz) [inline workshops/views.py]: …or `redirect(event)`, right? The latter is less popular but IMO a little bit clearer where you end up after the redirect.

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 minor issues and one/two challenges, and apart from them +1 (again, not going to merge until after #873 is fixed).

**Comment by narayanaditya95** [comment]
@pbanaszkiewicz Where do I claim my 1000 bonus virtual points?

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 good job, I'll merge soon into v1.7.

**Comment by pbanaszkiewicz** [comment]
@narayanaditya95 can you please rebase? Thanks!

**Comment by narayanaditya95** [comment]
Done
